File: recipes/claude/03-test-api-connectivity/typescript/test_connectivity.ts
import Anthropic from "@anthropic-ai/sdk";
import * as dotenv from "dotenv";

dotenv.config();

const apiKey = process.env.ANTHROPIC_API_KEY;
if (!apiKey) {
  console.error("Error: ANTHROPIC_API_KEY environment variable is missing.");
  process.exit(1);
}

const anthropic = new Anthropic({ apiKey });

async function probeModels(): Promise<boolean> {
  console.log("--- 1. Probing Models Endpoint ---");
  const start = performance.now();
  try {
    const list = await anthropic.models.list({ limit: 20 });
    const elapsedMs = performance.now() - start;
    console.log(`Status: OK (Retrieved models in ${elapsedMs.toFixed(2)} ms)`);
    console.log(`Total models returned: ${list.data.length}`);
    for (const model of list.data.slice(0, 5)) {
      console.log(`  - ${model.id}`);
    }
    return true;
  } catch (error: any) {
    if (error instanceof Anthropic.RateLimitError) {
      console.error(`Rate limited (HTTP 429): ${error.message}`);
    } else if (error instanceof Anthropic.InternalServerError) {
      console.error(`Anthropic Overloaded/Server Error (HTTP ${error.status}): ${error.message}`);
    } else if (error instanceof Anthropic.APIConnectionError) {
      console.error(`Connection failed: ${error.message}`);
    } else {
      console.error(`Unexpected API error: ${error.message}`);
    }
    return false;
  }
}

async function probeLatency(model = "claude-3-5-haiku-20241022"): Promise<boolean> {
  console.log(`\n--- 2. Measuring Latency Probe (${model}) ---`);
  const start = performance.now();
  try {
    const message = await anthropic.messages.create({
      model,
      max_tokens: 1,
      messages: [{ role: "user", content: "ping" }],
    });
    const elapsedMs = performance.now() - start;
    console.log(`Status: OK (Roundtrip latency: ${elapsedMs.toFixed(2)} ms)`);
    console.log(`Model: ${message.model}`);
    console.log(`Tokens used: prompt=${message.usage.input_tokens}, completion=${message.usage.output_tokens}`);
    return true;
  } catch (error: any) {
    if (error instanceof Anthropic.RateLimitError) {
      console.error(`Rate limited (HTTP 429): ${error.message}`);
    } else if (error instanceof Anthropic.InternalServerError) {
      console.error(`Server error (HTTP ${error.status}): ${error.message}`);
    } else if (error instanceof Anthropic.APIConnectionError) {
      console.error(`Connection failed: ${error.message}`);
    } else {
      console.error(`Unexpected API error: ${error.message}`);
    }
    return false;
  }
}

async function main() {
  const modelsOk = await probeModels();
  const latencyOk = await probeLatency();
  if (modelsOk && latencyOk) {
    console.log("\nAll connectivity and model probes passed.");
    process.exit(0);
  } else {
    console.log("\nConnectivity probe encountered failures.");
    process.exit(1);
  }
}

main();

File: recipes/claude/03-test-api-connectivity/python/test_connectivity.py
import os
import sys
import time
from dotenv import load_dotenv
import anthropic
from anthropic import Anthropic, RateLimitError, InternalServerError, APIConnectionError, APIStatusError

# Load environment variables
load_dotenv()

api_key = os.environ.get("ANTHROPIC_API_KEY")
if not api_key:
    print("Error: ANTHROPIC_API_KEY is not set.")
    sys.exit(1)

client = Anthropic(api_key=api_key)

def probe_models():
    """Query available Claude models via client.models.list()."""
    print("--- 1. Probing Models Endpoint ---")
    start = time.perf_counter()
    try:
        response = client.models.list(limit=20)
        elapsed_ms = (time.perf_counter() - start) * 1000
        print(f"Status: OK (Retrieved models in {elapsed_ms:.2f} ms)")
        model_ids = [m.id for m in response.data]
        print(f"Total models available: {len(model_ids)}")
        for mid in model_ids[:5]:
            print(f"  - {mid}")
        return True, elapsed_ms
    except RateLimitError as e:
        print(f"Rate limited (HTTP 429): {e.message}")
        return False, None
    except InternalServerError as e:
        print(f"Anthropic Overloaded/Internal error (HTTP {e.status_code}): {e.message}")
        return False, None
    except APIConnectionError as e:
        print(f"Network connection failed: {e}")
        return False, None
    except APIStatusError as e:
        print(f"API Error (HTTP {e.status_code}): {e.message}")
        return False, None

def probe_latency(model: str = "claude-3-5-haiku-20241022"):
    """Measure single-token ping latency via client.messages.create()."""
    print(f"\n--- 2. Measuring Latency Probe ({model}) ---")
    start = time.perf_counter()
    try:
        response = client.messages.create(
            model=model,
            max_tokens=1,
            messages=[{"role": "user", "content": "ping"}],
        )
        elapsed_ms = (time.perf_counter() - start) * 1000
        print(f"Status: OK (Roundtrip latency: {elapsed_ms:.2f} ms)")
        print(f"Model: {response.model}")
        print(f"Usage: prompt_tokens={response.usage.input_tokens}, completion_tokens={response.usage.output_tokens}")
        return True, elapsed_ms
    except RateLimitError as e:
        print(f"Rate limited (HTTP 429): {e.message}")
        return False, None
    except InternalServerError as e:
        print(f"Server error (HTTP {e.status_code}): {e.message}")
        return False, None
    except APIConnectionError as e:
        print(f"Connection failure: {e}")
        return False, None
    except APIStatusError as e:
        print(f"API Error (HTTP {e.status_code}): {e.message}")
        return False, None

if __name__ == "__main__":
    ok_models, _ = probe_models()
    ok_latency, _ = probe_latency()
    if ok_models and ok_latency:
        print("\nAll connectivity and model probes passed.")
        sys.exit(0)
    else:
        print("\nConnectivity probes encountered errors.")
        sys.exit(1)
